File: apps/android/app/src/main/java/com/smartble/core/utils/VersionMetadata.kt
package com.smartble.core.utils

import com.smartble.core.config.RELEASE_METADATA

/**
 * F027 版本元数据投影（PRD §5 F027 · R26 · DATA_MODEL「版本投影」）。
 *
 * uniapp `services/version-metadata.js` 的 Kotlin 锁定镜像：只读
 * com.smartble.core.config.RELEASE_METADATA（与 uniapp 产物出自同一确定性管线），
 * 无 IO / 网络。P009 关于页版本三态 = 运行时渠道成功 → 覆盖；失败 → 回退
 * Release Metadata 投影（verFallback）；元数据版本亦空 → "dev.unknown"。
 * P010 版本记录页 = [versionPageModel] 纯函数模型，不手写版本事实。
 */
object VersionMetadata {

    data class PlatformPublicStatus(
        val key: String,
        val name: String,
        val role: String,
        val capabilityStatus: String?,
        val releaseStatus: String,
    )

    data class CapabilityPublicStatus(
        val key: String,
        val name: String,
        val role: String,
        val capabilityStatus: String?,
        val releaseStatus: String,
        val reason: String?,
    )

    data class ReleaseRecord(
        val version: String,
        val tag: String,
        val channel: String,
        val status: String,
        val builtAt: String?,
        val commit: String?,
        val artifacts: List<Any?>,
    )

    data class PreviewRecord(
        val version: String,
        val label: String,
        val channel: String,
        val status: String,
    )

    data class Current(
        val version: String,
        val status: String,
        val channel: String,
        val channelLabel: String,
        val platforms: List<PlatformPublicStatus>,
        val limitations: List<String>,
        val hasReleaseTag: Boolean,
        val hasArtifacts: Boolean,
        val displayVersion: String,
    )

    data class History(
        val releases: List<ReleaseRecord>,
        val previews: List<PreviewRecord>,
    )

    data class PageModel(
        val current: Current,
        val history: History,
    )

    /** 仓库当前 Release Metadata（生成物，键序与 uniapp 产物一致）。 */
    fun releaseMetadata(): Map<String, Any?> = RELEASE_METADATA

    /** 元数据版本真值（根 VERSION 单源经生成管线投出）。 */
    fun productVersion(): String = str(RELEASE_METADATA["app_version"])

    /**
     * 版本串三态（正典 buildVersionString）：
     * 版本空 → "dev.unknown"；release 渠道 → v+sha（无 sha 裸版本）；
     * 其余渠道 → v-dev.sha（无 sha → v-dev.unknown）。
     */
    fun buildVersionString(version: String?, commit: String?, channel: String?): String {
        val v = version?.trim().orEmpty()
        if (v.isEmpty()) return "dev.unknown"

        val ch = (channel?.takeIf { it.isNotEmpty() } ?: "preview").lowercase()
        val sha = shortSha(commit)

        return if (ch == "release") {
            if (sha.isEmpty()) v else "$v+$sha"
        } else {
            if (sha.isEmpty()) "$v-dev.unknown" else "$v-dev.$sha"
        }
    }

    /** P009 运行时渠道失败时的回退基准（verFallback）。 */
    fun metadataVersionLabel(): String = buildVersionString(
        version = RELEASE_METADATA["app_version"] as? String,
        commit = RELEASE_METADATA["commit"] as? String,
        channel = RELEASE_METADATA["channel"] as? String,
    )

    /** 平台公开状态（顺序=正典 order 七键）。 */
    fun platformPublicStatuses(): List<PlatformPublicStatus> {
        val surfaces = surfaces()
        return listOf(
            "android", "wechat", "h5", "ios",
            "flutter_tauri_native", "peripheral", "observer",
        ).mapNotNull { key ->
            (surfaces[key] as? Map<*, *>)?.let { platformStatus(key, it) }
        }
    }

    /** 能力公开状态（ota BLOCKED / smart_hid PREVIEW 等如实展示）。 */
    fun capabilityPublicStatuses(): List<CapabilityPublicStatus> {
        val surfaces = surfaces()
        return listOf("ota", "smart_hid").mapNotNull { key ->
            (surfaces[key] as? Map<*, *>)?.let {
                CapabilityPublicStatus(
                    key = key,
                    name = str(it["name"]),
                    role = str(it["role"]),
                    capabilityStatus = it["capability_status"] as? String,
                    releaseStatus = str(it["release_status"]),
                    reason = it["reason"] as? String,
                )
            }
        }
    }

    /** 正典 platformDisplayStatus：REFERENCE 是角色直接展示；否则
     * capability → release → NOT_RELEASED 兜底。 */
    fun platformDisplayStatus(surface: Map<*, *>): String {
        if (surface["role"] == "REFERENCE") return "REFERENCE"
        val cap = str(surface["capability_status"])
        val rel = str(surface["release_status"])
        return when {
            cap.isNotEmpty() -> cap
            rel.isNotEmpty() -> rel
            else -> "NOT_RELEASED"
        }
    }

    /** P010 页面模型：纯函数，只消费 Release Metadata（可注入，便于测试）。 */
    fun versionPageModel(metadata: Map<String, Any?>? = null): PageModel {
        val meta = metadata ?: RELEASE_METADATA
        val version = str(meta["app_version"])
        val channel = ((meta["channel"] as? String)?.takeIf { it.isNotEmpty() } ?: "preview").lowercase()
        val status = (meta["overall_status"] as? String)?.takeIf { it.isNotEmpty() } ?: "PREVIEW"
        val surfaces = meta["public_surfaces"] as? Map<*, *> ?: emptyMap<Any?, Any?>()

        val platforms = listOf("android", "wechat", "h5", "ios").mapNotNull { key ->
            (surfaces[key] as? Map<*, *>)?.let { platformStatus(key, it) }
        }

        val limitations = (meta["known_limitations"] as? List<*>)?.map { it.toString() } ?: emptyList()
        val artifacts = (meta["artifacts"] as? List<*>) ?: emptyList<Any?>()
        val hasArtifacts = artifacts.isNotEmpty()
        val releaseTag = str(meta["release_tag"])
        val hasReleaseTag = releaseTag.isNotEmpty()

        val releases = if (hasReleaseTag && hasArtifacts) {
            listOf(
                ReleaseRecord(
                    version = version,
                    tag = releaseTag,
                    channel = "release",
                    status = "VERIFIED",
                    builtAt = meta["built_at"] as? String,
                    commit = meta["commit"] as? String,
                    artifacts = artifacts,
                )
            )
        } else {
            emptyList()
        }

        val previews = if (version.isNotEmpty() && (channel == "preview" || channel == "dev")) {
            listOf(
                PreviewRecord(
                    version = version,
                    label = "$version Preview",
                    channel = channel,
                    status = status,
                )
            )
        } else {
            emptyList()
        }

        return PageModel(
            current = Current(
                version = version,
                status = status,
                channel = channel,
                channelLabel = if (channel == "release") "Release" else "Preview",
                platforms = platforms,
                limitations = limitations,
                hasReleaseTag = hasReleaseTag,
                hasArtifacts = hasArtifacts,
                displayVersion = buildVersionString(
                    version = version,
                    commit = meta["commit"] as? String,
                    channel = channel,
                ),
            ),
            history = History(releases = releases, previews = previews),
        )
    }

    private fun platformStatus(key: String, surface: Map<*, *>): PlatformPublicStatus =
        PlatformPublicStatus(
            key = key,
            name = str(surface["name"]),
            role = str(surface["role"]),
            capabilityStatus = surface["capability_status"] as? String,
            releaseStatus = str(surface["release_status"]),
        )

    private fun surfaces(): Map<*, *> =
        RELEASE_METADATA["public_surfaces"] as? Map<*, *> ?: emptyMap<Any?, Any?>()

    private fun str(v: Any?): String = v?.toString()?.trim() ?: ""

    private fun shortSha(commit: String?): String {
        val t = commit?.trim().orEmpty()
        return if (t.isEmpty()) "" else t.take(7)
    }
}

File: apps/flutter/lib/core/utils/version_metadata.dart
/// F027 版本元数据投影（PRD §5 F027 · R26 · DATA_MODEL「版本投影」）。
///
/// uniapp `services/version-metadata.js` 的 Dart 锁定镜像：只读
/// `core/config/release_metadata_generated.dart`（同一确定性管线产物），
/// 无 IO / 网络。P009 关于页版本三态 = 运行时渠道成功 → 覆盖；失败 →
/// 回退 Release Metadata 投影（verFallback）；元数据版本亦空 → `dev.unknown`。
/// P010 版本记录页 = 本投影的纯函数模型（getVersionPageModel），不手写版本事实。
library;

import '../config/release_metadata_generated.dart';

String _shortSha(String? commit) {
  if (commit == null || commit.trim().isEmpty) return '';
  return commit.trim().substring(0, 7);
}

String _str(dynamic v) => v == null ? '' : v.toString().trim();

/// 仓库当前 Release Metadata（生成物，键序与 uniapp 产物一致）。
Map<String, dynamic> getReleaseMetadata() => kReleaseMetadata;

/// 元数据版本真值（根 VERSION 单源经生成管线投出）。
String getProductVersion() => _str(kReleaseMetadata['app_version']);

/// 版本串三态（正典 buildVersionString）：
/// 版本空 → `dev.unknown`；release 渠道 → `v+sha`（无 sha 裸版本）；
/// 其余渠道 → `v-dev.sha`（无 sha → `v-dev.unknown`）。
String buildVersionString({String? version, String? commit, String? channel}) {
  final v = _str(version);
  if (v.isEmpty) return 'dev.unknown';

  final rawCh = channel ?? '';
  final ch = (rawCh.isEmpty ? 'preview' : rawCh).toLowerCase();
  final sha = _shortSha(commit);

  if (ch == 'release') {
    return sha.isEmpty ? v : '$v+$sha';
  }
  return sha.isEmpty ? '$v-dev.unknown' : '$v-dev.$sha';
}

/// P009 运行时渠道失败时的回退基准（verFallback）。
String metadataVersionLabel() => buildVersionString(
      version: getProductVersion(),
      commit: _strOrNull(kReleaseMetadata['commit']),
      channel: _strOrNull(kReleaseMetadata['channel']),
    );

String? _strOrNull(dynamic v) => v?.toString();

class PlatformPublicStatus {
  const PlatformPublicStatus({
    required this.key,
    required this.name,
    required this.role,
    required this.capabilityStatus,
    required this.releaseStatus,
  });

  final String key;
  final String name;
  final String role;
  final String? capabilityStatus;
  final String releaseStatus;
}

class CapabilityPublicStatus {
  const CapabilityPublicStatus({
    required this.key,
    required this.name,
    required this.role,
    required this.capabilityStatus,
    required this.releaseStatus,
    this.reason,
  });

  final String key;
  final String name;
  final String role;
  final String? capabilityStatus;
  final String releaseStatus;
  final String? reason;
}

/// 平台公开状态（P009 推广区外链卡等消费；顺序=正典 order 七键）。
List<PlatformPublicStatus> getPlatformPublicStatuses() {
  const order = [
    'android',
    'h5',
    'ios',
    'flutter_tauri_native',
    'peripheral',
    'observer',
  ];
  final surfaces = _surfaces();
  return [
    for (final key in order)
      if (surfaces[key] != null)
        PlatformPublicStatus(
          key: key,
          name: _str(surfaces[key]!['name']),
          role: _str(surfaces[key]!['role']),
          capabilityStatus: _strOrNull(surfaces[key]!['capability_status']),
          releaseStatus: _str(surfaces[key]!['release_status']),
        ),
  ];
}

/// 能力公开状态（ota BLOCKED / smart_hid PREVIEW 等如实展示）。
List<CapabilityPublicStatus> getCapabilityPublicStatuses() {
  final surfaces = _surfaces();
  return [
    for (final key in ['ota', 'smart_hid'])
      if (surfaces[key] != null)
        CapabilityPublicStatus(
          key: key,
          name: _str(surfaces[key]!['name']),
          role: _str(surfaces[key]!['role']),
          capabilityStatus: _strOrNull(surfaces[key]!['capability_status']),
          releaseStatus: _str(surfaces[key]!['release_status']),
          reason: _strOrNull(surfaces[key]!['reason']),
        ),
  ];
}

Map<String, dynamic> _surfaces() {
  final s = kReleaseMetadata['public_surfaces'];
  if (s is Map<String, dynamic>) return s;
  return const <String, dynamic>{};
}

/// 正典 platformDisplayStatus：REFERENCE 是角色直接展示；否则
/// capability → release → NOT_RELEASED 兜底。
String platformDisplayStatus(Map<String, dynamic> surface) {
  if (surface['role'] == 'REFERENCE') return 'REFERENCE';
  return _str(surface['capability_status']).isNotEmpty
      ? _str(surface['capability_status'])
      : _str(surface['release_status']).isNotEmpty
          ? _str(surface['release_status'])
          : 'NOT_RELEASED';
}

class ReleaseRecord {
  const ReleaseRecord({
    required this.version,
    required this.tag,
    required this.channel,
    required this.status,
    this.builtAt,
    this.commit,
    required this.artifacts,
  });

  final String version;
  final String tag;
  final String channel;
  final String status;
  final String? builtAt;
  final String? commit;
  final List<dynamic> artifacts;
}

class PreviewRecord {
  const PreviewRecord({
    required this.version,
    required this.label,
    required this.channel,
    required this.status,
  });

  final String version;
  final String label;
  final String channel;
  final String status;
}

class VersionPageCurrent {
  const VersionPageCurrent({
    required this.version,
    required this.status,
    required this.channel,
    required this.channelLabel,
    required this.platforms,
    required this.limitations,
    required this.hasReleaseTag,
    required this.hasArtifacts,
    required this.displayVersion,
  });

  final String version;
  final String status;
  final String channel;
  final String channelLabel;
  final List<PlatformPublicStatus> platforms;
  final List<String> limitations;
  final bool hasReleaseTag;
  final bool hasArtifacts;
  final String displayVersion;
}

class VersionPageHistory {
  const VersionPageHistory({required this.releases, required this.previews});

  final List<ReleaseRecord> releases;
  final List<PreviewRecord> previews;
}

class VersionPageModel {
  const VersionPageModel({required this.current, required this.history});

  final VersionPageCurrent current;
  final VersionPageHistory history;
}

/// P010 页面模型：纯函数，只消费 Release Metadata（可注入，便于测试）。
VersionPageModel getVersionPageModel([Map<String, dynamic>? metadata]) {
  final meta = metadata ?? kReleaseMetadata;
  final version = _str(meta['app_version']);
  final rawChannel = meta['channel']?.toString() ?? '';
  final channel = (rawChannel.isEmpty ? 'preview' : rawChannel).toLowerCase();
  final rawStatus = meta['overall_status']?.toString() ?? '';
  final status = rawStatus.isEmpty ? 'PREVIEW' : rawStatus;
  final surfaces = _surfacesOf(meta);

  final platforms = [
    for (final key in ['android', 'h5', 'ios'])
      if (surfaces[key] != null)
        PlatformPublicStatus(
          key: key,
          name: _str(surfaces[key]!['name']),
          role: _str(surfaces[key]!['role']),
          capabilityStatus: _strOrNull(surfaces[key]!['capability_status']),
          releaseStatus: _str(surfaces[key]!['release_status']),
        ),
  ];

  final rawLimitations = meta['known_limitations'];
  final limitations = rawLimitations is List
      ? [for (final x in rawLimitations) x.toString()]
      : <String>[];

  final rawArtifacts = meta['artifacts'];
  final artifacts = rawArtifacts is List ? rawArtifacts : const <dynamic>[];
  final hasArtifacts = artifacts.isNotEmpty;
  final releaseTag = _str(meta['release_tag']);
  final hasReleaseTag = releaseTag.isNotEmpty;

  final releases = <ReleaseRecord>[];
  if (hasReleaseTag && hasArtifacts) {
    releases.add(ReleaseRecord(
      version: version,
      tag: releaseTag,
      channel: 'release',
      status: 'VERIFIED',
      builtAt: _strOrNull(meta['built_at']),
      commit: _strOrNull(meta['commit']),
      artifacts: artifacts,
    ));
  }

  final previews = <PreviewRecord>[];
  if (version.isNotEmpty && (channel == 'preview' || channel == 'dev')) {
    previews.add(PreviewRecord(
      version: version,
      label: '$version Preview',
      channel: channel,
      status: status,
    ));
  }

  return VersionPageModel(
    current: VersionPageCurrent(
      version: version,
      status: status,
      channel: channel,
      channelLabel: channel == 'release' ? 'Release' : 'Preview',
      platforms: platforms,
      limitations: limitations,
      hasReleaseTag: hasReleaseTag,
      hasArtifacts: hasArtifacts,
      displayVersion: buildVersionString(
        version: version,
        commit: _strOrNull(meta['commit']),
        channel: channel,
      ),
    ),
    history: VersionPageHistory(releases: releases, previews: previews),
  );
}

Map<String, dynamic> _surfacesOf(Map<String, dynamic> meta) {
  final s = meta['public_surfaces'];
  if (s is Map<String, dynamic>) return s;
  return const <String, dynamic>{};
}
